source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,507 @@
+
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            {
+
+              // Get old stresses from compositional fields
+              SymmetricTensor<2,dim> stress_old;
+              for (unsigned int j=0; j < SymmetricTensor<2,dim>::n_independent_components; ++j)
+                {
+                  stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)] = in.composition[i][j];
+                }
+
+              // Calculate the rotated stresses
+              // Rotation (vorticity) tensor (equation 25 in Moresi et al., 2003, J. Comp. Phys.)
+              const Tensor<2,dim> rotation = 0.5 * ( old_velocity_gradients[i] - transpose(old_velocity_gradients[i]) );
+
+              // Recalculate average values of viscosity, elastic shear modulus and viscoelastic (effective) viscosity
+              const std::vector<double> composition = in.composition[i];
+              const std::vector<double> volume_fractions = compute_volume_fractions(composition);
+              const double vis_avg = average_value ( volume_fractions, viscosities, viscosity_averaging);
+              const double esm_avg = average_value ( volume_fractions, elastic_shear_moduli, viscosity_averaging);
+              const double vev_avg = ( vis_avg * dte ) / ( dte + ( vis_avg / esm_avg ) );
+
+              // Calculate new stress terms (see equation 29 in Moresi et al., 2003, J. Comp. Phys.)
+              SymmetricTensor<2,dim> stress_new = ( 2. * vev_avg * deviator(in.strain_rate[i]) ) +
+                                                  ( ( vev_avg / ( esm_avg * dte ) ) * stress_old ) +
+                                                  ( ( vev_avg / esm_avg ) *
+                                                    ( symmetrize(rotation * Tensor<2,dim>(stress_old) ) - symmetrize(Tensor<2,dim>(stress_old) * rotation) ) );
+
+              // Stress averaging scheme to account for difference betweed fixed elastic time step
+              // and numerical time step (see equation 32 in Moresi et al., 2003, J. Comp. Phys.)
+              const double dt = this->get_timestep();
+              if (use_fixed_elastic_time_step == true && use_stress_averaging == true)
+                {
+                  stress_new = ( ( 1. - ( dt / dte ) ) * stress_old ) + ( ( dt / dte ) * stress_new ) ;
+                }
+
+              // Fill reaction terms
+              for (unsigned int j = 0; j < SymmetricTensor<2,dim>::n_independent_components ; ++j)
+                out.reaction_terms[i][j] = -stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)]
+                                           + stress_new[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)];
+
+              // Fill elastic outputs (see equation 30 in Moresi et al., 2003, J. Comp. Phys.)
+              if (force)
+                {
+                  force->rhs_e[i] = -1. * ( ( vev_avg / ( esm_avg * dte  ) ) * stress_old );
```

Comment:
I decided that the best option is to enforce setting `enable_elasticity` to true through an exception. If the user would like to not have any elastic effects, they can modify the shear modulus so that the viscosity and viscoelastic viscosity are effectively equal.